java/src/main/java/io/cucumber/gherkin/GherkinTokenMatcher.java

```diff
@@ -20,30 +24,39 @@
 final class GherkinTokenMatcher implements TokenMatcher {
 
     private static final Pattern LANGUAGE_PATTERN = Pattern.compile("^#\\s*language\\s*:\\s*([a-zA-Z\\-_]+)\\s*$");
-    private final GherkinDialectProvider dialectProvider;
-    private GherkinDialect currentDialect;
+    private static final String DEFAULT_LANGUAGE = "en";
+    private final String defaultLanguage;
+    // Expect at most two languages, the default language and one other
+    private final Map<String, KeywordMatcher> activeKeywordMatchers = new HashMap<>(2);
+    private String currentLanguage;
+    private KeywordMatcher currentKeywordMatcher;
     private String activeDocStringSeparator = null;
     private int indentToRemove = 0;
 
-    GherkinTokenMatcher(GherkinDialectProvider dialectProvider) {
-        this.dialectProvider = dialectProvider;
+    GherkinTokenMatcher(String defaultLanguage) {
+        this.defaultLanguage = defaultLanguage;
         reset();
     }
 
     GherkinTokenMatcher() {
-        this(new GherkinDialectProvider());
-    }
-
-    GherkinTokenMatcher(String defaultDialectName) {
-        this(new GherkinDialectProvider(defaultDialectName));
+        this(DEFAULT_LANGUAGE);
     }
 
     @Override
     public void reset() {
         // TODO performance: reset() is called once in the constructor and once for each file (Parser.parse()). It could be called only once, but there is no measurable impact with the profiler
         activeDocStringSeparator = null;
         indentToRemove = 0;
-        currentDialect = dialectProvider.getDefaultDialect();
+        setLanguageMatched(defaultLanguage, null);
+    }
+
+    private void setLanguageMatched(String language, Location location) {
+        currentLanguage = language;
```

Comment:
If the language didn't change, the `activeKeywordMatchers` loopkup can be avoided.